Model/ProductProtection.php

```diff
@@ -7,94 +7,227 @@
     ) {
-        $this->cart = $cart;
-        $this->productFactory = $productFactory;
+        $this->logger = $logger;
+        $this->quoteRepository = $quoteRepository;
+        $this->itemFactory = $itemFactory;
+        $this->itemOptionFactory = $itemOptionFactory;
+        $this->checkoutSession = $checkoutSession;
+        $this->integration = $integration;
+        $this->storeManager = $storeManager;
+        $this->productRepository = $productRepository;
     }
 
     /**
-     * Add product protection to cart
+     * Upsert product protection in cart
      *
-     * @param int $quantity
-     * @param string $productId
-     * @param string $planId
-     * @param int $price
-     * @param int $term
-     * @param string $coverageType
-     * @param string $token = null
-     * @param float $listPrice = null
-     * @param string $orderOfferPlanId = null
+     * @param int|null $quantity
+     * @param string|null $cartItemId
+     * @param string|null $productId
+     * @param string|null $planId
+     * @param int|null $price
+     * @param int|null $term
+     * @param string|null $coverageType
+     * @param string|null $leadToken
+     * @param string|null $listPrice
+     * @param string|null $orderOfferPlanId
      * @return void
+     * @throws NoSuchEntityException
      */
-    public function add(int $quantity, string $productId, string $planId, int $price, int $term, string $coverageType, string $leadToken = null, float $listPrice = null, string $orderOfferPlanId = null): void {
-        $product = $this->productFactory->create();
-        $product->load($product->getIdBySku(Extend::WARRANTY_PRODUCT_SKU));
+    public function upsert(
+        int $quantity = null,
+        string $cartItemId = null,
+        string $productId = null,
+        string $planId = null,
+        int $price = null,
+        int $term = null,
+        string $coverageType = null,
+        string $leadToken = null,
+        string $listPrice = null,
+        string $orderOfferPlanId = null
+    ): void {
+        try {
+            $product = $this->productRepository->get(Extend::WARRANTY_PRODUCT_SKU);
+            $quote = $this->checkoutSession->getQuote();
+            $quoteId = $quote->getId();
 
-        $params = array();
-        $options = array();
-        $params['qty'] = $quantity;
-        $params['product'] = $product->getId();
-        $product->setData('extend_plan_price', ($price/100));
+            if ($price === 0) {
+                throw new Exception('Cannot add/update product protection with a price of 0');
+            }
 
+            // if quantity is 0, remove the item from the quote
+            if ($quantity === 0 && isset($cartItemId)) {
+                $quote->removeItem($cartItemId);
+                $this->quoteRepository->save($quote);
+                return;
+            }
+
+            if (isset($cartItemId)) {
+                $item = $this->checkoutSession->getQuote()->getItemById($cartItemId);
```

Comment:
I forget how it works in PHP but `$item` doesn't need to be declared before this does it?

Reply:
No, since if statements do not have their own scope it is still the function scope